src/gpu/vulkan/SDL_gpu_vulkan.c

```diff
@@ -8186,6 +8186,124 @@ static void VULKAN_UploadToBuffer(
+    vulkanTextureSlice = VULKAN_INTERNAL_SDLToVulkanTextureSlice(&textureRegion->textureSlice);
+    VulkanResourceAccessInfo resourceAccessInfo;
+
+    resourceAccessInfo.stageMask = VK_PIPELINE_STAGE_TRANSFER_BIT;
+    resourceAccessInfo.accessMask = VK_ACCESS_TRANSFER_WRITE_BIT;
+    resourceAccessInfo.imageLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+
+    VULKAN_INTERNAL_BufferMemoryBarrier(
+        renderer,
+        vulkanCommandBuffer,
+        &resourceAccessInfo,
+        SDL_TRUE,
+        transferBufferContainer->activeBufferHandle->vulkanBuffer
+    );
+
+    resourceAccessInfo.stageMask = VK_PIPELINE_STAGE_TRANSFER_BIT;
+    resourceAccessInfo.accessMask = VK_ACCESS_TRANSFER_READ_BIT;
+    resourceAccessInfo.imageLayout = VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL;
+
+    VULKAN_INTERNAL_ImageMemoryBarrier(
+        renderer,
+        vulkanCommandBuffer,
+        &resourceAccessInfo,
+        SDL_TRUE,
+        vulkanTextureSlice
+    );
+
+    imageCopy.imageExtent.width = textureRegion->w;
+    imageCopy.imageExtent.height = textureRegion->h;
+    imageCopy.imageExtent.depth = textureRegion->d;
+    imageCopy.imageOffset.x = textureRegion->x;
+    imageCopy.imageOffset.y = textureRegion->y;
+    imageCopy.imageOffset.z = textureRegion->z;
+    imageCopy.imageSubresource.aspectMask = vulkanTextureSlice->parent->aspectFlags;
+    imageCopy.imageSubresource.baseArrayLayer = textureRegion->textureSlice.layer;
+    imageCopy.imageSubresource.layerCount = 1;
+    imageCopy.imageSubresource.mipLevel = textureRegion->textureSlice.mipLevel;
+    imageCopy.bufferOffset = copyParams->bufferOffset;
+    imageCopy.bufferRowLength = copyParams->bufferStride;
+    imageCopy.bufferImageHeight = copyParams->bufferImageHeight;
+
+    renderer->vkCmdCopyImageToBuffer(
+        vulkanCommandBuffer->commandBuffer,
+        vulkanTextureSlice->parent->image,
+        VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL,
+        transferBufferContainer->activeBufferHandle->vulkanBuffer->buffer,
+        1,
+        &imageCopy
+    );
+
+    VULKAN_INTERNAL_TrackBuffer(renderer, vulkanCommandBuffer, transferBufferContainer->activeBufferHandle->vulkanBuffer);
+    VULKAN_INTERNAL_TrackTextureSlice(renderer, vulkanCommandBuffer, vulkanTextureSlice);
+}
+
+static void VULKAN_DownloadFromBuffer(
+    SDL_GpuCommandBuffer *commandBuffer,
+    SDL_GpuBuffer *gpuBuffer,
+    SDL_GpuTransferBuffer *transferBuffer,
+    SDL_GpuBufferCopy *copyParams
+) {
+    VulkanCommandBuffer *vulkanCommandBuffer = (VulkanCommandBuffer*) commandBuffer;
+    VulkanRenderer *renderer = vulkanCommandBuffer->renderer;
+    VulkanBufferContainer *gpuBufferContainer = (VulkanBufferContainer*) gpuBuffer;
+    VulkanBufferContainer *transferBufferContainer = (VulkanBufferContainer*) transferBuffer;
+    VkBufferCopy bufferCopy;
+    VulkanResourceAccessInfo resourceAccessInfo;
+
+    resourceAccessInfo.stageMask = VK_PIPELINE_STAGE_TRANSFER_BIT;
+    resourceAccessInfo.accessMask = VK_ACCESS_SHADER_WRITE_BIT;
```

Comment:
I guess we need "shader" read/write in order to do DMA via vkCmdCopyBuffer, not transfer read/write? Because looking at vkCmdCopyBuffer, it seems to only require that the src buffer have TRANSFER_SRC_BIT usage and TRANSFER_DST_BIT usage for the dst buffer, which suggests to me that the copy can be a transfer and not a shader operation?

Reply:
https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#copies

All copy commands are treated as “transfer” operations for the purposes of synchronization barriers.

Reply:
Nice catch, I think autocomplete got me here